src/tests/tdycore_tests.h

```diff
@@ -9,46 +9,108 @@
+// registered to be called on program exit.
+static int _run_selected_tests(int argc, char **argv,
+                               void (*setup)(int argc, char **argv),
                                int num_tests,
                                const struct CMUnitTest tests[num_tests],
+                               void (*breakdown)(void),
                                int nproc) {
-  if (strcasecmp(command, "count") == 0) { // asked for # of tests
-    fprintf(stdout, "%d\n", num_tests);
-    return 0;
-  } else if (strcasecmp(command, "nproc") == 0) { // asked for # of procs
-    fprintf(stdout, "%d\n", nproc);
-    return 0;
-  } else { // asked for a specific test index
-    // Try to interpret the argument as an index for the desired test.
-    char *endptr;
-    long index = strtol(command, &endptr, 10);
-    if (*endptr == '\0') { // got a valid index!
-      if ((index < 0) || (index >= num_tests)) {
-        fprintf(stderr, "Invalid test index: %ld (must be in [0, %d])\n",
-            index, num_tests);
-        return 1;
+
+  if (argc == 1) {
+    if (setup != NULL) {
+      setup(argc, argv);
+    }
+    fprintf(stdout, "Usage:\n");
+    fprintf(stdout, "%s count [arg1 [...]]  \treports the number of available unit tests\n", argv[0]);
+    fprintf(stdout, "%s nproc [arg1 [...]]  \treports the # of MPI procs supported\n", argv[0]);
+    fprintf(stdout, "%s <index> [arg1 [...]]\truns the unit test with the (0-based) index\n", argv[0]);
+    return 1;
+  } else {
+    const char* command = (const char*)argv[1];
+    if (strcasecmp(command, "count") == 0) { // asked for # of tests
+      fprintf(stdout, "%d\n", num_tests);
+      return 0;
+    } else if (strcasecmp(command, "nproc") == 0) { // asked for # of procs
+      fprintf(stdout, "%d\n", nproc);
+      return 0;
+    } else { // asked for a specific test index
+      // Try to interpret the argument as an index for the desired test.
+      char *endptr;
+      long index = strtol(command, &endptr, 10);
+      if (*endptr == '\0') { // got a valid index!
+        if ((index < 0) || (index >= num_tests)) {
+          if (num_tests == 1) {
+            fprintf(stderr, "Invalid test index: %ld (must be 0)\n", index);
+          } else {
+            fprintf(stderr, "Invalid test index: %ld (must be between 0 and %d)\n",
+                    index, num_tests-1);
+          }
+          return 1;
+        } else {
+          // We have a valid test index. If we have been given a setup function,
+          // construct our own argc and argv with the index argument removed and
+          // call the function with them.
+          int my_argc = -1;
+          char** my_argv = NULL;
+          if (setup != NULL) {
+            my_argc = argc - 1;
+            my_argv = malloc(my_argc * sizeof(char*));
+            my_argv[0] = _strdup(argv[0]);
+            for (int i = 1; i < my_argc; ++i) {
+              my_argv[i] = _strdup(argv[i+1]);
```

Comment:
Do we need to dup all these strings? I assume the concern is that the test could modify the values of our original `argv[i]`, but maybe this is harmless if we aren't using `argv[i]` afterward. (Not sure if that's true.)

Reply:
I was just trying to be paranoid about it, since it doesn't cost a lot of resources. :-)